htp #19094 fix double pull in PoolConductor

rkuhn opened the pull request:
lacking a test case, just for discussion

rkuhn committed:
#19094 fix double pull in PoolConductor

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolConductor.scala:
```diff
@@ -140,15 +140,15 @@ private object PoolConductor {
+          val wasBlocked = nextSlot == -1
           nextSlot = bestSlot()
-          if (tryPull) tryPullCtx()
+          if (wasBlocked && nextSlot != -1) pull(ctxIn)
```
I am still not fully understanding what this intends to do. I get it that it checks for the transition of nextSlot == -1 to nextSlot != -1, but I don't get why this is a useful event to decide when to pull `ctxIn`?

rkuhn replied:
The slot selector only blocks if the currently best slot back-pressures or if there is no available slot—as soon as one becomes available it should pull again because it will be able to emit the resulting request as well.

drewhk replied:
ah right, this is the slot port. Maybe name `nextSlot != -1` to `nowUnblocked` and add a comment:

    if (wasBlocked && nowUnblocked) pull(ctxIn) // Get next request context

Not to nitpick, but this is mostly alien code to me so any help is nice to have :)

rkuhn replied:
good point, will do

rkuhn committed:
fix actor leak in streams

The leak could happen if using sub-materializer and the substream
terminates during its initial runBatch() already.

rkuhn committed:
#19417 fix empty.head error in WebSocket termination

also fixes related issues in HTTP Server and OutgoingConnection and
addresses one flaw in Websocket.BypassRouter

rkuhn commented:
fixes #19094 

fixes #19417

drewhk commented on akka-http-core/src/test/scala/akka/http/impl/engine/ws/BypassRouterSpec.scala:
```diff
@@ -0,0 +1,49 @@
+      val (future, t) = Http().singleWebsocketRequest(
+        WebsocketRequest("ws://127.0.0.1:8080"),
+        Flow.fromSinkAndSourceMat(
+          Sink.ignore.mapMaterializedValue(_.map(_ ⇒ "hallo")),
```
I would check the number of messages received

drewhk replied:
Btw, a Sink.count would be useful.

ktoso replied:
Yeah, comes up sometimes - ticketify

rkuhn replied:
#19422 (will use `fold` here)

drewhk commented:
LGTM

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -75,8 +75,8 @@ private[http] object OutgoingConnectionBlueprint {
-        case _ ⇒
-          true
+        case (seq, _) ⇒
+          seq.nonEmpty
```
I see

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/ws/Websocket.scala:
```diff
@@ -172,13 +172,12 @@ private[http] object Websocket {
         }
       })
-      val pullIn = () ⇒ pull(in)
+      val pullIn = () ⇒ tryPull(in)
```
Hm, try pull is `if (!closed(in))`, I thought we'll want to pull `if (!hasBeenPulled(in))` - why try here?

rkuhn replied:
This case was not a double-pull, it was a pull-when-closed that happens because we unconditionally execute the pull even when the stage was completed while the emit was running.

ktoso replied:
Cool, I see

ktoso commented:
LGTM, had one question for my better understanding of the fix
